fluka_sobp_source/source_sampler.f

```diff
@@ -0,0 +1,635 @@
+         WRITE(LUNOUT,*) 'SOBP SOURCE ZPOS fixed to', ZBEAM
+         WRITE(LUNOUT,*) 'SOBP USER PARAM LIST', WHASOU
+
+*        Fluka run happens in a temporary directory,
+*        created in same level as input file
+*        sobp.dat is not copied there,
+*        We reach one level up to get it via ../sobp.dat
+         CALL READSOBP ( '../sobp.dat', ENERGY,DE,
+     $            XPOS, YPOS, FWHMX, FWHMY, PART, NOCOLUMNS, NWEIGHT )
+
+
+*        In case of problem with reading sobp.dat file
+         IF ( (NOCOLUMNS .LE. ZERZER) .OR. (NWEIGHT .LE. ZERZER)) THEN
+            NOMORE = 1
+            RETURN
+         ENDIF
+
+*        In sobp.dat file energy is saved in GeV/amu, while
+*        in Fluka we need it in not per amu, but simply in GeV
+*        To achieve this we multiply energy by mass number A
+*        Fluka doesn't have any consistent method of calculating
+*        mass number (number of nucleons) for simple particles
+*        (such as protons and alphas) and heavy ions
+*        we use a method IBARCH to get baryonic charge which
+*        reduces to mass number for particles of interest
+         ENERGY = ENERGY * IBARCH(IONID)
+
+*        First parameter in SOURCE indicates position of virtual source.
+*        If the number is present and non-zero we expect point-like source
+         IF ( WHASOU(1) .NE. 0.0D0 ) THEN
+            WRITE(LUNOUT,*) 'SOBP POINT-LIKE VIRTUAL SOURCE'
+            LPNTSRC = .TRUE.
+         ELSE
+            WRITE(LUNOUT,*) 'SOBP PARALLEL VIRTUAL SOURCE'
+            LPNTSRC = .FALSE.
+         END IF
+
+      END IF
+*  |  End of first call initializations.
+*  +-------------------------------------------------------------------*
+
+*  +-------------------------------------------------------------------*
+*  |  sample beamlet collection :
+
+
+*     Get a 64-bit pseudo random number in the interval [0.D+00,1.D+00),
+*     1 being not included
+      RAN = FLRNDM(111)
+
+*     Now an integer random number, from the 1 to NWEIGHT,
+*     needed to select random line from sobp config file
+*     http://infohost.nmt.edu/tcc/help/lang/fortran/scaling.html
+*     If you want an integer between i and j inclusive
+*     use int(rand(0)*(j+1-i))+i
+      NRAN = INT(RAN * NWEIGHT) + 1
+
+*     check if random number selected properly
+      IF ((NRAN .GT. NWEIGHT) .OR. (NRAN .LT. 1)) THEN
+         WRITE(LUNOUT,*) 'SOBP SOURCE ERROR. NRAN, RAN:', NRAN, RAN
+         NOMORE = 3
+         RETURN
+      END IF
+
+
+*  +-------------------------------------------------------------------*
+*  Push one source particle to the stack. Note that you could as well
+*  push many but this way we reserve a maximum amount of space in the
+*  stack for the secondaries to be generated
+* Npflka is the stack counter: of course any time source is called it
+* must be =0
+      NPFLKA = NPFLKA + 1
+* Wt is the weight of the particle
+      WTFLK  (NPFLKA) = PART(NRAN)    ! set new weight
```

Comment:
not sure if I like the style with two whitespace between array name and array index. Is there a special point why they do that?

Reply:
I have no clue, inherited from Fluka example, I will gladly fix it.